acme/api/middleware.go

```diff
@@ -176,9 +188,18 @@ func (h *Handler) extractJWK(next nextHTTP) nextHTTP {
 		}
 		ctx = context.WithValue(ctx, jwkContextKey, jwk)
 
+		if len(jwk.KeyID) == 0 {
+			kid, err := jwk.Thumbprint(crypto.SHA256)
+			if err != nil {
+				api.WriteError(w, acme.ServerInternalErr(errors.Wrap(err, "error generating jwk thumbprint")))
+				return
+			}
+			jwk.KeyID = base64.RawURLEncoding.EncodeToString(kid)
+		}
```

Comment:
Is this a question, or a comment, or an observation?

Reply:
An observation, I know that some `kid` values are URLs, so just making sure you're taking this into account if it can have any effect here.

Reply:
Ah, yes so you're correct in that some kid values are urls. However context is king. The 'kid' protected header of the JWS (body of the request) must always be the account url if it is defined. If it is not defined, then the 'jwk' protected header must contain the public JWK, and we use the kid of the jwk to look up the account. That is the case in this method. We are looking up the account using a JWK, specifically the 'kid' of the JWK.

Reply:
The 'jwk' and 'kid' protected headers are mutually exclusive and one of them must be defined.

Reply:
New question as to whether we should always re-calculate the keyid. otherwise a bad actor could pass in a JWK with incorrect key-id to load account information for an account they don't own. They wouldn't be able to make any other requests though as the signature in the JWS would not pass the sig check.

Reply:
I replaced using the jwk.KeyID by always recalculating the keyID when looking up and storing the account information.